src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+            }
+        }
+        else
+        {
+            RemoveVariablesFromContext(new List<string> { variableName }, contextInformation);
+        }
+
+        return evaluationResult.CanEvaluate;
+    }
+
+    private static bool ProcessAssignment(AssignmentExpressionSyntax assignmentExpression, ContextInformation contextInformation, HashSet<string> dirtyVariables, bool displayDiagnostics)
+    {
+        var evaluationResult = PreEvaluate(assignmentExpression, contextInformation, null, displayDiagnostics);
+        var variableName = (assignmentExpression.Left.TrimParenthesis() as IdentifierNameSyntax).Identifier.ValueText;
+
+        if (evaluationResult.CanEvaluate)
+        {
+            var variableValue = evaluationResult.SyntaxNodeToVariableValueMap.Values.First();
+            var generateNewContext = IsCompoundExpression(assignmentExpression.Kind()) || evaluationResult.SyntaxNodeToVariableValueMap.Count() > 1;
+
+            if (generateNewContext)
+            {
+                var expressionAnalysisContext = GenerateNewExpressionAnalysisContext(assignmentExpression, evaluationResult.SyntaxNodeToVariableValueMap);
+                variableValue = new VariableValue(expressionAnalysisContext, new List<Location>());
+            }
+
+            dirtyVariables.Add(variableName);
+            contextInformation.VariableValues[variableName] = variableValue;
+            var originalExpression = variableValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (contextInformation.BuilderNodesProcessed.Add(originalExpression))
+            {
+                contextInformation.VariableValuesList.Add(variableValue);
+            }
+
+            if (assignmentExpression.Parent is ConditionalExpressionSyntax)
+            {
+                RemoveVariablesFromContext(new List<string>() { variableName }, contextInformation);
+            }
+        }
+        else
+        {
+            RemoveVariablesFromContext(new List<string>() { variableName }, contextInformation);
+        }
+
+        return evaluationResult.CanEvaluate;
+    }
+
+    private static void RemoveVariablesFromContext(IEnumerable<string> variablesToRemove, ContextInformation contextInformation)
+    {
+        var variableValues = contextInformation.VariableValues;
+        var variableValuesList = contextInformation.VariableValuesList;
+        var builderNodesProcessed = contextInformation.BuilderNodesProcessed;
+
+        foreach (var variable in variablesToRemove)
+        {
+            var isValidVariable = variableValues.TryGetValueSafe(variable, out var previousValue);
+
+            if (!isValidVariable)
+            {
+                continue;
+            }
+
+            var originalExpression = previousValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (builderNodesProcessed.Add(originalExpression))
+            {
+                variableValuesList.Add(previousValue);
```

Comment:
In what case variableValuesList does not contain previousValue already?

Reply:
If I have the following code snippet. 
var x = exp1;
var y = x;
var z = x;
When I reach the end of the method/block context, I must remove x, y, and z. Due to our optimization, x, y, and z all "share" the same VariableValue and this VariableValue contains the ExpressionAnalysisContext for exp1. Hence, When I'm removing x, y, and z from the context, I don't want to add the same VariableValue 3 times. Without this explicit check here, it will be added 3 times and the same diagnostic may be shown multiple times at a location